src/datasets/utils/download_manager.py

```diff
@@ -225,21 +228,36 @@ def iter_archive(self, path):
             Generator yielding tuple (path_within_archive, file_obj).
             File-Obj are opened in byte mode (io.BufferedReader)
         """
-        with open(path, "rb") as f:
-            stream = tarfile.open(fileobj=f, mode="r|*")
-            for tarinfo in stream:
-                file_path = tarinfo.name
-                if not tarinfo.isreg():
+        extension = Path(path).suffix
```

Comment:
I think cached zip or tar archives don't have the extension at the end of their filenames. Also for streaming we don't always know the filename either.

You can take a look at the `_get_extraction_protocol` function in `utils/streaming_download_manager.py`. It first checks the extension and then fallbacks to using the magic number of the file to guess the compression type